File: benchmarks/python/birdfcn.py
import numpy as np

def birdfcn(x: np.ndarray) -> np.ndarray:
    """
    Computes the value of the Bird function.

    Parameters:
    -----------
    x (numpy.ndarray): A matrix of size M-by-2.

    Returns:
    --------
    numpy.ndarray: A vetor SCORES of size M-by-1 in which each row contains 
    the function value for the corresponding row of X.
    
    Raises
    ------
    AssertionError
        If the input array has a number of columns different from 2.

    Author: Mazhar Ansari Ardeh
    Please forward any comments or bug reports to mazhar.ansari.ardeh at
    Google's e-mail service or feel free to kindly modify the repository.
    """

    n = x.shape[1]
    assert n == 2, 'Bird function is only defined on a 2D space.'
    X = x[:, 0]
    Y = x[:, 1]

    scores = np.sin(X) * np.exp((1 - np.cos(Y)) ** 2) + \
             np.cos(Y) * np.exp((1 - np.sin(X)) ** 2) + \
             (X - Y) ** 2

    return scores

File: src/matlab/birdfcn.m
% Computes the value of the Bird function.
% SCORES = BIRDFCN(X) computes the value of the Bird
% function at point X. BIRDFCN accepts a matrix of size M-by-2 and
% returns a vetor SCORES of size M-by-1 in which each row contains the
% function value for the corresponding row of X.
% For more information, please visit:
% benchmarkfcns.info/doc/birdfcn
%
% Author: Mazhar Ansari Ardeh
% Please forward any comments or bug reports to mazhar.ansari.ardeh at
% Google's e-mail service or feel free to kindly modify the repository.
function scores = birdfcn(x)

    n = size(x, 2);
    assert(n == 2, 'Bird function is only defined on a 2D space.')
    X = x(:, 1);
    Y = x(:, 2);

    scores = sin(X) .* exp((1 - cos(Y)).^2) + ...
        cos(Y) .* exp((1 - sin(X)) .^ 2) + ...
        (X - Y) .^ 2;
end
